File: front-end/src/SideComponents/bstagiaire.py
from flask import Flask, jsonify, request, make_response
from flask_cors import CORS
from pymongo import MongoClient
from bson import ObjectId
import logging

app = Flask(__name__)
CORS(app, resources={r"/stagiaire": {"origins": "*"}})  # Enable CORS for the /stagiaire route

client = MongoClient("mongodb://localhost:27017/")
db = client["system_facial"]
collection = db["stagiaire"]

logging.basicConfig(level=logging.INFO)

@app.route('/')
def index():
    return 'Welcome to the backend!'

@app.route('/stagiaire', methods=['GET', 'OPTIONS'])
def get_stagiaire():
    if request.method == "OPTIONS":  
        return _build_cors_prelight_response()
    else:
        try:
            stagiaire_list = []
            for stagiaire in collection.find():
                stagiaire_list.append({
                    "_id": str(stagiaire["_id"]),
                    "nom": stagiaire["nom"],
                    "prenom": stagiaire["prenom"],
                    "date_naissance": stagiaire["date_naissance"],
                    "genre": stagiaire["genre"]
                })
            return jsonify(stagiaire_list)
        except Exception as e:
            error_msg = f"Error fetching stagiaires: {e}"
            logging.error(error_msg)
            return jsonify({"error": error_msg}), 500

def _build_cors_prelight_response():
    response = make_response()
    response.headers.add("Access-Control-Allow-Origin", "*")
    response.headers.add("Access-Control-Allow-Headers", "*")
    response.headers.add("Access-Control-Allow-Methods", "*")
    return response

if __name__ == '__main__':
    app.run(debug=True, port=5000)

File: backend/Backend-Sidebare/bstagiaire.js
const express = require('express');
const cors = require('cors');
const MongoClient = require('mongodb').MongoClient;
const logger = require('morgan');

const app = express();
const port = 5000;

app.use(express.json());
app.use(express.urlencoded({ extended: true }));
app.use(logger('dev'));
app.use(cors()); // Enable CORS for all routes

const mongoURI = 'mongodb://localhost:27017/';
const dbName = 'system_facial';
const stagiaireCollectionName = 'stagiaire';

let db, stagiaireCollection;

MongoClient.connect(mongoURI, { useUnifiedTopology: true })
    .then(client => {
        db = client.db(dbName);
        stagiaireCollection = db.collection(stagiaireCollectionName);
        console.log('Connected to MongoDB');
    })
    .catch(error => console.error('Error connecting to MongoDB:', error));

app.get('/', (req, res) => {
    res.send('Welcome to the backend!');
});

app.get('/stagiaire', async (req, res) => {
    try {
        const { nom_groub } = req.query;
        let query = {};
        if (nom_groub) {
            query.nom_groub = nom_groub;
        }
        const stagiaireList = await stagiaireCollection.find(query).toArray();
        const formattedStagiaireList = stagiaireList.map(stagiaire => ({
            _id: stagiaire._id.toString(),
            nom: stagiaire.nom,
            prenom: stagiaire.prenom,
            date_naissance: stagiaire.date_naissance,
            genre: stagiaire.genre,
            nom_groub: stagiaire.nom_groub
        }));
        res.json(formattedStagiaireList);
    } catch (error) {
        const errorMsg = `Error fetching stagiaires: ${error}`;
        console.error(errorMsg);
        res.status(500).json({ error: errorMsg });
    }
});

app.get('/api/groups', async (req, res) => {
    try {
        const groups = await stagiaireCollection.distinct('nom_groub');
        res.json(groups.map(group => ({ nom_groub: group })));
    } catch (error) {
        const errorMsg = `Error fetching groups: ${error}`;
        console.error(errorMsg);
        res.status(500).json({ error: errorMsg });
    }
});

app.listen(port, () => {
    console.log(`Server is running on port ${port}`);
});
